### PR TITLE
Round to the nearest integer when computing missile coordinates

### PR DESCRIPTION
**Pull request by StephenCWills**
Fixes the distortion in "missile space" described in #2852. By rounding the result of this computation to the nearest integer, it effectively identifies the coordinates of the tile whose center is closest to the screen coordinates.

This also fixes the 4-tile bug where ranged monsters will choose to stand and shoot in a position that allows the player to hit them but not the other way around. Missiles should now travel along exactly the same path in both directions.

https://user-images.githubusercontent.com/9203145/206085029-0bf95d14-4d90-4c0b-b704-f14476fb74d1.mp4

### CODE REVIEW

**Comment by StephenCWills** [comment]
Argh, right. This would cause the demo to fail.

**Review by ephphatha — DISMISSED** [review]
I haven't tested this but it matches what I was experimenting with earlier, nice that it's all isolated in that one transformation 🙂

File: Source/engine/displacement.hpp [hunk 1]
```diff
@@ -155,11 +155,16 @@ struct DisplacementOf {
...
 	 */
 	constexpr DisplacementOf<DeltaT> screenToMissile() const
 	{
-		return -screenToWorld();
+		static_assert(std::is_signed<DeltaT>::value, "DeltaT must be signed for transformations involving a rotation");
+		DeltaT xNumerator = 2 * deltaY + deltaX;
+		DeltaT yNumerator = 2 * deltaY - deltaX;
+		DeltaT xOffset = (xNumerator >= 0) ? 32 : -32;
+		DeltaT yOffset = (yNumerator >= 0) ? 32 : -32;
+		return { (xNumerator + xOffset) / 64, (2 * deltaY - deltaX + yOffset) / 64 };
```
[the remark this answers is not in the document]
Comment (StephenCWills) [reply Source/engine/displacement.hpp]: Yes, thanks for catching that.

**Comment by glebm** [comment]
Might be good to record a new timedemo anyway because the old one fails when run outside the test for unknown reasons.

If you record a new timedemo, can you please also apply the fix here:
https://github.com/diasurgical/devilutionX/blob/c8c3f9366c2208d871e8a784595f0161dafd6e35/Source/objects.h#L63-L69

**Comment by AJenbo** [comment]
We should probably figure out why that is :)

**Comment by StephenCWills** [comment]
@glebm In what way does the timedemo fail? I ran it just now without any issue.

**Comment by AJenbo** [comment]
Here is a new demo recording that goes through more of the available events in the two levels from the same save as the previous demo. The demo passes when played visually... but fails when used in the test. The issue appears to be that the game never ends and then tries to run past the end at which results in "Demo queue empty"

[demo_0.zip](https://github.com/diasurgical/devilutionX/files/10181446/demo_0.zip)

I'm not sure what broke, but I see that the demo does not have a `SDL_QUIT` event?

**Comment by AJenbo** [comment]
There are still a few cases where the headless demos will fail.

- Switching equipment that changes the heros graphics
- Anything that triggers time based messages (like shrines), as they interferer with pressing `esc`
- Ending the recording using the mouse

**Comment by glebm** [comment]
The new timedemo passes for me finally :tada: 

INFO: 3688 frames, 16.65 seconds: 221.5 fps
INFO: Timedemo: Same outcome as inital run. :)

**Pull request merged**